workflow/commit-messages.md

```diff
@@ -0,0 +1,43 @@
+---
+title: Commit guidelines
+description: Some general guidelines to help you in writing good commits for collaboration and clarity.
+---
+
+# Commit Guidelines
+
+As a general guideline you can follow the idea that _a single commit should enclose a single problem/feature that has been fixed/added_.
+
+This way of organizing your changes to the codebase makes is easy for reviewers and people who try to understand the history of a project to determine what has been happening in the project.
+
+Here a both a good and a bad example of of commit message.
+
+**Bad:**
+
+```
+fixes bug
+```
+
+**Good:**
+
+```
+Fix bug where database gets deleted on POST request
+```
+
+## Breaking up large tickets
+
+If your feature/ticket includes mutliple things that you have to complete in order to submit a pull request, these should all be seperate commit messages, according to the general guideline above. This will make it easy for the reviewer to understand what exactly has happened.
```

Comment:
> these should all be seperate commit messages

Do you actually mean seperate commit messages, or multi-line commit messages?

Reply:
As far as I understand the intention was for it to be separate commits (with proper messages).

I agree that it's usually tricky. We don't often have a perfect plan and errorless solution in mind that nicely maps in to a clean, logically split commit history. We implement a part of solution, commit it, maybe discover an error later, get some review comment. Keeping commit history clean with always valuable "feature" commits would require constant rebases with a sole purpose of having a clean history - history we will rarely look at.

If the PR itself is small enough (fixes one bug, adds one feature) it can theoretically work in history as one commit. Even if developed in couple of steps it's the final result that counts. When reviewing I rarely look at individual commits. It's the whole PR diff that I review.

While I agree that having meaningful commits with good commit messages is valuable, I wouldn't go out of the way cleaning up history, rebasing review comment fixes into past individual commits, etc.

Reply:
I am a bit concerned with this document that it will make people feel like they *have to* spend a lot of effort tidying commits, and I do agree that this is not worth spending a lot of time on.

Having said that, I do think this is a *little* blaze:

> Keeping commit history clean with always valuable "feature" commits would require constant rebases with a sole purpose of having a clean history - history we will rarely look at.

Clean history is definitely useful. While it's true that the vast majority of history probably won't be inspected, it's also true that people very often do inspect history. I definitely do. And I also link to commits. I tell the community in bug reports e.g. "this feature was added in this commit", or I quickly skim the commits between two dates to see what happened. Or sometimes I cherry-pick commits to quickly grab functionality from other projects. Creating neat, atomic commits that describe individual features that are added are definitely not a waste of time.

So the middle ground I think it would be good to achieve with this document is to outline what would be the ideal way to create commits, but making it clear that if you find it difficult it is not a hard requirement. I *never* want PRs to be blocked because of the format of commit messages - @b-m-f please could you add something to the document to make this clear?

It would certainly be helpful if all our devs would gradually learn how to quickly and efficiently manage Git commits such that it's second nature and doesn't take them much time, but this upskilling will always be a gradual process.

## Tricks for keeping commits tidy

Having said all that, personally, I do my best to rebase commits quite regularly. It's become so second nature that I don't think it slows me down in any significant way. A few simple tricks help.

1. (I think everyone does this), frequently *rebase* (don't merge) your work on top of the latest version of the branch you're merging into. Do this at every opportunity, and ideally just before the PR is merged.
2. Heavy use of `git commit --amend --no-edit`. If you do new work that could simply fit into the previous commit, just do `git add {changed filed}` followed by `git commit --amend --no-edit`
3. Heavy use of interactive rebasing (@tbille showed me this trick, I didn't know it before). Since this is fairly involved, I'll explain it below:

### Interactive rebasing

Let's say:

1. You do some work in `users.js` and you commit it with "Improve user logic"
2. You do some work in `README.md` and you commit it with "Explain user logic better in README"
3. Now you realise you need another fix to your `users.js` work. I want it to be added to my "Improve user logic" commit, so I do this:

``` bash
$ git add users.js
$ git commit -m '... to be rebased'
$ git rebase -i HEAD^^^  # Interactively rebase the last 3 commits
```

This will open an editor with this contents:

```
pick 239d4c3 Improve user logic
pick adc7c21 Explain user logic better in README
pick fd4f81b ... to be rebased
```

Now if I reorder my commits to move line 3 to line 2, and replace "pick" with "squash" (or simply "s"), then it will achieve what I want:

```
pick 239d4c3 Improve user logic
squash fd4f81b ... to be rebased
pick adc7c21 Explain user logic better in README
```

Save and exit, and then it will open another editor to ask you what you want the new commit message for the merged commit to be:

```
# This is a combination of 2 commits.
# This is the 1st commit message:

Improve user logic

# This is the commit message #2:

... to be rebased
```

But in this case I probably just want to delete the second message. Save and exist, and voila:

``` bash
$ git log
commit b87706ed594c983841857b51923e499988760d41 (HEAD -> master)
Author: Robin Winslow <robin@robinwinslow.co.uk>
Date:   Fri Dec 7 11:05:54 2018 +0000

    Explain user logic better in README

commit da1e9d47073cdc3bf86d3658b2e020c7e37292c0
Author: Robin Winslow <robin@robinwinslow.co.uk>
Date:   Fri Dec 7 11:05:40 2018 +0000

    Improve user logic
```

This process might seem complicated, but once you've done it a couple of times, you'll see that it actually doesn't take long. It probably takes me usually about 15-20 seconds.

There are of course some cases where commits can't be neatly re-ordered (although fewer than you'd think). In these cases, after you perform your interactive rebase, you'll get a conflict which you'll have to resolve in the normal way. You'll quickly get a feel for which commits can be reordered and which can't, and if you think it's going to be too much trouble, just don't bother starting. Or `git rebase --abort`.

Reply:
`git rebase -i` is great once you get used to it. I do use it quite regularly on a larger PRs to turn the long list of commits into something more managable (mostly by squash and fixup - as reordering is tricky as @nottrobin pointed out).

@nottrobin That is a great write down here. It would be good to turn it into a document itself or a part of one about git/commit/github process.

Reply:
Thanks @bartaz. I realise that at this point I've basically written a blog post. I should probably publish it =)

Reply:
@nottrobin You should!

Reply:
Thank @b-m-f. I've tidied the Wiki post a bit and renamed it:

https://github.com/canonical-webteam/practices/wiki/Tips-on-managing-commit-history-in-Git

I think it would be good to link to it from this document.

Reply:
I'm turning this into a blog post - the draft is here: https://docs.google.com/document/d/1gPBV9NBPCwVDcEo6mRZSF--WmDgoBa5hL3Uzzc0aZTI/edit

Please feel free to jump in and make any suggestions.

Reply:
@b-m-f I think the language here is still ambiguous.

Reply:
>Clean history is definitely useful. While it's true that the vast majority of history probably won't be inspected, it's also true that people very often do inspect history. I definitely do. And I also link to commits. I tell the community in bug reports e.g. "this feature was added in this commit", or I quickly skim the commits between two dates to see what happened. Or sometimes I cherry-pick commits to quickly grab functionality from other projects. Creating neat, atomic commits that describe individual features that are added is definitely not a waste of time.

Absolutely, but I think we're conflating clean history on master with clean history in a PR, which is what @bartaz was talking about.

Reply:
I published the blog post: https://blog.ubuntu.com/2018/12/12/tricks-for-keeping-a-tidy-git-commit-history

Reply:
> Absolutely, but I think we're conflating clean history on master with clean history in a PR, which is what @bartaz was talking about.

For all our repositories on GitHub (except maybe build.snapcraft.io, which we should maybe look into), history in a PR becomes history in master, because we "merge" rather than "squash" (In GitHub, even if you "squash", by default each individual commit message is still included in the overall commit message of the squash commit).

@b-m-f it may be worth just making it clear that caring about the commit history in your PR is largely irrelevant if the repository you're working on squashes commits when PRs are merged. With a note that, in this scenario, you should try to be as detailed as possible in your squash commit.

Reply:
@squidsoup I have added another sentence for the squashing. Do you think this makes it  less ambiguous?